File: speechiq_frontend_flutter/lib/features/analysis/models/analysis_result.dart
import 'dart:math';

// ── Public model ──────────────────────────────────────────────────────────────

class AnalysisResult {
  final int overallScore;
  final MetricScore fluency;
  final MetricScore pacing;
  final MetricScore clarity;
  final MetricScore confidence;
  final MetricScore vocabulary;
  final MetricScore pronunciation;

  // API feedback — list of prioritised, specific improvement tips
  final List<String> feedback;

  // Real API fields
  final String transcript;
  final List<TranscriptSegment> segments;
  final double duration;
  final PauseMetrics pauseMetrics;
  final SpeechMetrics speechMetrics;
  final List<TimelineSegment> timeline;

  AnalysisResult({
    required this.overallScore,
    required this.fluency,
    required this.pacing,
    required this.clarity,
    required this.confidence,
    required this.vocabulary,
    required this.pronunciation,
    List<String>? feedback,
    this.transcript = '',
    this.segments = const [],
    this.duration = 0,
    PauseMetrics? pauseMetrics,
    SpeechMetrics? speechMetrics,
    this.timeline = const [],
  })  : feedback = feedback ?? const [],
        pauseMetrics = pauseMetrics ?? PauseMetrics.empty(),
        speechMetrics = speechMetrics ?? SpeechMetrics.empty();

  /// Parses both shapes the API can return:
  ///   Full response  → {"file": "...", "analysis": { ... }}
  ///   Bare analysis  → { "transcript": ..., "scores": ..., ... }
  factory AnalysisResult.fromJson(Map<String, dynamic> json) {
    final a = json.containsKey('analysis')
        ? json['analysis'] as Map<String, dynamic>
        : json;

    final rawScores = a['scores'] as Map<String, dynamic>? ?? {};

    int clamp(num v) => v.toDouble().clamp(0.0, 100.0).round();
    int potential(int score) => min(score + 12, 100);
    double toDouble(dynamic v) => (v as num? ?? 0).toDouble();

    final fluencyScore      = clamp(toDouble(rawScores['fluency']));
    final pacingScore       = clamp(toDouble(rawScores['pacing']));
    final clarityScore      = clamp(toDouble(rawScores['clarity']));
    final confidenceScore   = clamp(toDouble(rawScores['confidence']));
    final vocabularyScore   = clamp(toDouble(rawScores['vocabulary']));
    final pronunciationScore= clamp(toDouble(rawScores['pronunciation']));
    final overallScore      = clamp(toDouble(rawScores['overall']));

    final pausesMap = a['pauses'] as Map<String, dynamic>? ?? {};
    final pm = PauseMetrics.fromJson(
        pausesMap['metrics'] as Map<String, dynamic>? ?? {});
    final sm = SpeechMetrics.fromJson(
        a['speech'] as Map<String, dynamic>? ?? {});

    // Prefer the API's feedback list; fall back to local builder
    final rawFeedback = a['feedback'];
    final List<String> feedbackList = rawFeedback is List && rawFeedback.isNotEmpty
        ? rawFeedback.map((e) => e.toString()).toList()
        : _buildFeedback(fluencyScore, pacingScore, clarityScore, pm, sm);

    return AnalysisResult(
      overallScore:  overallScore,
      fluency:       MetricScore(score: fluencyScore,      potential: potential(fluencyScore)),
      pacing:        MetricScore(score: pacingScore,       potential: potential(pacingScore)),
      clarity:       MetricScore(score: clarityScore,      potential: potential(clarityScore)),
      confidence:    MetricScore(score: confidenceScore,   potential: potential(confidenceScore)),
      vocabulary:    MetricScore(score: vocabularyScore,   potential: potential(vocabularyScore)),
      pronunciation: MetricScore(score: pronunciationScore,potential: potential(pronunciationScore)),
      feedback:      feedbackList,
      transcript:    a['transcript'] as String? ?? '',
      segments:      (a['segments'] as List<dynamic>? ?? [])
          .map((e) => TranscriptSegment.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
      duration:      toDouble(a['duration']),
      pauseMetrics:  pm,
      speechMetrics: sm,
      timeline:      (a['timeline'] as List<dynamic>? ?? [])
          .map((e) => TimelineSegment.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
    );
  }

  static List<String> _buildFeedback(
    int fluency, int pacing, int clarity,
    PauseMetrics pm, SpeechMetrics sm,
  ) {
    final items = <String>[];
    items.add(
      'You spoke at ${sm.speechRateWpm.round()} WPM'
      '${pm.pauseCount > 0 ? ' with ${pm.pauseCount} pauses (avg ${pm.avgPauseDuration.toStringAsFixed(1)}s).' : '.'}',
    );
    if (fluency < 50) {
      items.add('Fluency needs attention — focus on reducing long silences.');
    } else if (pacing < 50) {
      items.add('Work on keeping a steadier speaking pace throughout.');
    } else if (clarity < 50) {
      items.add('Articulation of word endings could be sharper.');
    } else {
      items.add('Good overall delivery — keep practising to build consistency.');
    }
    return items;
  }

  /// Silence segments as 0–1 waveform highlight ranges.
  List<Map<String, double>> get silenceHighlights {
    if (duration <= 0) return [];
    return timeline
        .where((s) => s.type == 'silence')
        .map((s) => {
              'start': (s.start / duration).clamp(0.0, 1.0),
              'end':   (s.end   / duration).clamp(0.0, 1.0),
            })
        .toList();
  }

  /// The lowest-scoring implemented metric (for coaching card).
  String get worstMetric {
    final scores = {
      'Fluency':    fluency.score,
      'Pacing':     pacing.score,
      'Clarity':    clarity.score,
      'Confidence': confidence.score,
      'Vocabulary': vocabulary.score,
    };
    return scores.entries
        .reduce((a, b) => a.value <= b.value ? a : b)
        .key;
  }
}

// ── Sub-models ────────────────────────────────────────────────────────────────

class MetricScore {
  final int score;
  final int potential;
  const MetricScore({required this.score, required this.potential});
}

class TranscriptSegment {
  final int id;
  final double start;
  final double end;
  final String text;

  const TranscriptSegment({
    required this.id,
    required this.start,
    required this.end,
    required this.text,
  });

  factory TranscriptSegment.fromJson(Map<String, dynamic> j) =>
      TranscriptSegment(
        id:    (j['id']    as num? ?? 0).toInt(),
        start: (j['start'] as num? ?? 0).toDouble(),
        end:   (j['end']   as num? ?? 0).toDouble(),
        text:  j['text']   as String? ?? '',
      );
}

class PauseMetrics {
  final double totalSpeechTime;
  final double totalSilenceTime;
  final int pauseCount;
  final double avgPauseDuration;
  final int longPauseCount;   // pauses >= 1 second
  final double pauseRate;     // pauses per minute of speech

  const PauseMetrics({
    required this.totalSpeechTime,
    required this.totalSilenceTime,
    required this.pauseCount,
    required this.avgPauseDuration,
    this.longPauseCount = 0,
    this.pauseRate = 0.0,
  });

  factory PauseMetrics.empty() => const PauseMetrics(
      totalSpeechTime: 0, totalSilenceTime: 0,
      pauseCount: 0, avgPauseDuration: 0);

  factory PauseMetrics.fromJson(Map<String, dynamic> j) => PauseMetrics(
        totalSpeechTime:  (j['total_speech_time']  as num? ?? 0).toDouble(),
        totalSilenceTime: (j['total_silence_time'] as num? ?? 0).toDouble(),
        pauseCount:       (j['pause_count']        as num? ?? 0).toInt(),
        avgPauseDuration: (j['avg_pause_duration'] as num? ?? 0).toDouble(),
        longPauseCount:   (j['long_pause_count']   as num? ?? 0).toInt(),
        pauseRate:        (j['pause_rate']         as num? ?? 0).toDouble(),
      );
}

class SpeechMetrics {
  final int wordCount;
  final int uniqueWordCount;
  final double speechRateWpm;
  final double wordsPerSecond;
  final double speechRatio;
  final double averageWordsPerSegment;
  final Map<String, int> fillers;

  const SpeechMetrics({
    required this.wordCount,
    required this.uniqueWordCount,
    required this.speechRateWpm,
    required this.wordsPerSecond,
    required this.speechRatio,
    required this.averageWordsPerSegment,
    required this.fillers,
  });

  int get totalFillers => fillers.values.fold(0, (s, v) => s + v);
  double get fillerRatePer100 =>
      wordCount > 0 ? totalFillers / wordCount * 100 : 0;
  double get typeTokenRatio =>
      wordCount > 0 ? uniqueWordCount / wordCount : 0;

  factory SpeechMetrics.empty() => const SpeechMetrics(
      wordCount: 0, uniqueWordCount: 0, speechRateWpm: 0,
      wordsPerSecond: 0, speechRatio: 0,
      averageWordsPerSegment: 0, fillers: {});

  factory SpeechMetrics.fromJson(Map<String, dynamic> j) => SpeechMetrics(
        wordCount:              (j['word_count']            as num? ?? 0).toInt(),
        uniqueWordCount:        (j['unique_word_count']     as num? ?? 0).toInt(),
        speechRateWpm:          (j['speech_rate_wpm']       as num? ?? 0).toDouble(),
        wordsPerSecond:         (j['words_per_second']      as num? ?? 0).toDouble(),
        speechRatio:            (j['speech_ratio']          as num? ?? 0).toDouble(),
        averageWordsPerSegment: (j['avg_words_per_segment'] as num? ?? 0).toDouble(),
        fillers: (j['fillers'] as Map? ?? {}).cast<String, int>(),
      );
}

class TimelineSegment {
  final double start;
  final double end;
  final String type; // 'speech' | 'silence'

  const TimelineSegment({
    required this.start,
    required this.end,
    required this.type,
  });

  factory TimelineSegment.fromJson(Map<String, dynamic> j) => TimelineSegment(
        start: (j['start'] as num).toDouble(),
        end:   (j['end']   as num).toDouble(),
        type:   j['type']  as String,
      );
}

File: speechiq_frontend_react/src/app/lib/analysisResult.ts
// Pure data types + parsing for the FastAPI /analyze response.
// Ported from speechiq_frontend_flutter/lib/features/analysis/models/analysis_result.dart
// No React/DOM imports on purpose — keep this testable in isolation.

export interface MetricScore {
  score: number;
  potential: number;
}

export interface TranscriptSegment {
  id: number;
  start: number;
  end: number;
  text: string;
}

export interface PauseMetrics {
  totalSpeechTime: number;
  totalSilenceTime: number;
  pauseCount: number;
  avgPauseDuration: number;
  longPauseCount: number;
  pauseRate: number;
}

export interface SpeechStats {
  wordCount: number;
  uniqueWordCount: number;
  speechRateWpm: number;
  wordsPerSecond: number;
  speechRatio: number;
  averageWordsPerSegment: number;
  fillers: Record<string, number>;
}

export interface TimelineSegment {
  start: number;
  end: number;
  type: string; // 'speech' | 'silence'
}

export interface IdealRange {
  min: number;
  max: number;
  unit: string;
}

/** One concrete flagged moment/pattern from the backend scorer. Timestamped
 *  events (start/end present) can be replayed in the UI; pattern-level ones
 *  (fillers, overused words) can't. */
export interface PenaltyEventData {
  description: string;
  start: number | null;
  end: number | null;
  points: number;
}

/** Per-dimension drill-down payload from scores.breakdown. `potential` is
 *  null when the API predates real potential computation (fallback heuristic
 *  applies instead). */
export interface DimensionBreakdown {
  potential: number | null;
  strengths: string[];
  penaltyEvents: PenaltyEventData[];
}

/** Keys match the backend's SpeechScorer.ideal_ranges() exactly. */
export type IdealRangeKey =
  | "speaking_rate"
  | "pause_rate"
  | "avg_pause_duration"
  | "avg_words_per_segment"
  | "speech_ratio"
  | "filler_rate_per_100_words"
  | "type_token_ratio";

/**
 * Fallback only — used when there's no real API result (e.g. MetricBreakdown's
 * standalone/demo usage). The backend (scoring/scorer.py: SpeechScorer.ideal_ranges)
 * is the single source of truth; these mirror its current values so the fallback
 * doesn't look wildly different, but are NOT re-derived from research independently.
 */
export const DEFAULT_IDEAL_RANGES: Record<IdealRangeKey, IdealRange> = {
  speaking_rate: { min: 125, max: 170, unit: "WPM" },
  pause_rate: { min: 0, max: 6, unit: "pauses/min" },
  avg_pause_duration: { min: 0.3, max: 1.2, unit: "seconds" },
  avg_words_per_segment: { min: 5, max: 15, unit: "words" },
  speech_ratio: { min: 70, max: 85, unit: "%" },
  filler_rate_per_100_words: { min: 0, max: 2, unit: "per 100 words" },
  type_token_ratio: { min: 60, max: 80, unit: "%" },
};

export interface AnalysisResult {
  overallScore: number;
  overallPotential: number;
  fluency: MetricScore;
  pacing: MetricScore;
  clarity: MetricScore;
  confidence: MetricScore;
  vocabulary: MetricScore;
  pronunciation: MetricScore;
  feedback: string[];
  transcript: string;
  segments: TranscriptSegment[];
  duration: number;
  pauseMetrics: PauseMetrics;
  speechMetrics: SpeechStats;
  timeline: TimelineSegment[];
  idealRanges: Record<IdealRangeKey, IdealRange>;
  /** Per-dimension strengths + penalty events, keyed by display category. */
  breakdowns: Partial<Record<MetricCategory, DimensionBreakdown>>;
}

export type MetricCategory = "Fluency" | "Pacing" | "Clarity" | "Confidence" | "Vocabulary";

function toNumber(v: unknown): number {
  return typeof v === "number" && !Number.isNaN(v) ? v : 0;
}

function clampScore(v: unknown): number {
  return Math.round(Math.min(100, Math.max(0, toNumber(v))));
}

function potentialFor(score: number): number {
  return Math.min(score + 12, 100);
}

function parsePauseMetrics(j: Record<string, unknown>): PauseMetrics {
  return {
    totalSpeechTime: toNumber(j["total_speech_time"]),
    totalSilenceTime: toNumber(j["total_silence_time"]),
    pauseCount: toNumber(j["pause_count"]),
    avgPauseDuration: toNumber(j["avg_pause_duration"]),
    longPauseCount: toNumber(j["long_pause_count"]),
    pauseRate: toNumber(j["pause_rate"]),
  };
}

function parseSpeechStats(j: Record<string, unknown>): SpeechStats {
  const fillers = j["fillers"];
  return {
    wordCount: toNumber(j["word_count"]),
    uniqueWordCount: toNumber(j["unique_word_count"]),
    speechRateWpm: toNumber(j["speech_rate_wpm"]),
    wordsPerSecond: toNumber(j["words_per_second"]),
    speechRatio: toNumber(j["speech_ratio"]),
    averageWordsPerSegment: toNumber(j["avg_words_per_segment"]),
    fillers: fillers && typeof fillers === "object" ? (fillers as Record<string, number>) : {},
  };
}

function parseSegments(arr: unknown): TranscriptSegment[] {
  if (!Array.isArray(arr)) return [];
  return arr.map((e) => {
    const j = (e ?? {}) as Record<string, unknown>;
    return {
      id: toNumber(j["id"]),
      start: toNumber(j["start"]),
      end: toNumber(j["end"]),
      text: typeof j["text"] === "string" ? (j["text"] as string) : "",
    };
  });
}

const BREAKDOWN_KEY_TO_CATEGORY: Record<string, MetricCategory> = {
  fluency: "Fluency",
  pacing: "Pacing",
  clarity: "Clarity",
  confidence: "Confidence",
  vocabulary: "Vocabulary",
};

/** Parses scores.breakdown into per-dimension potential/strengths/penalty events. */
function parseBreakdowns(j: unknown): Partial<Record<MetricCategory, DimensionBreakdown>> {
  const raw = (j ?? {}) as Record<string, unknown>;
  const result: Partial<Record<MetricCategory, DimensionBreakdown>> = {};
  for (const [key, category] of Object.entries(BREAKDOWN_KEY_TO_CATEGORY)) {
    const entry = raw[key];
    if (!entry || typeof entry !== "object") continue;
    const e = entry as Record<string, unknown>;
    const rawEvents = Array.isArray(e["penalty_events"]) ? (e["penalty_events"] as unknown[]) : [];
    result[category] = {
      potential: typeof e["potential"] === "number" ? clampScore(e["potential"]) : null,
      strengths: Array.isArray(e["strengths"]) ? (e["strengths"] as unknown[]).map(String) : [],
      penaltyEvents: rawEvents.map((ev) => {
        const p = (ev ?? {}) as Record<string, unknown>;
        return {
          description: typeof p["description"] === "string" ? (p["description"] as string) : "",
          start: typeof p["start"] === "number" ? (p["start"] as number) : null,
          end: typeof p["end"] === "number" ? (p["end"] as number) : null,
          points: toNumber(p["points"]),
        };
      }),
    };
  }
  return result;
}

/** Reads scores.ideal_ranges from the API, falling back per-key to
 *  DEFAULT_IDEAL_RANGES for any key the backend didn't send (e.g. an older
 *  API version) rather than leaving it undefined. */
function parseIdealRanges(j: unknown): Record<IdealRangeKey, IdealRange> {
  const raw = (j ?? {}) as Record<string, unknown>;
  const result = { ...DEFAULT_IDEAL_RANGES };
  for (const key of Object.keys(DEFAULT_IDEAL_RANGES) as IdealRangeKey[]) {
    const entry = raw[key];
    if (entry && typeof entry === "object") {
      const e = entry as Record<string, unknown>;
      result[key] = {
        min: toNumber(e["min"]),
        max: toNumber(e["max"]),
        unit: typeof e["unit"] === "string" ? (e["unit"] as string) : DEFAULT_IDEAL_RANGES[key].unit,
      };
    }
  }
  return result;
}

function parseTimeline(arr: unknown): TimelineSegment[] {
  if (!Array.isArray(arr)) return [];
  return arr.map((e) => {
    const j = (e ?? {}) as Record<string, unknown>;
    return {
      start: toNumber(j["start"]),
      end: toNumber(j["end"]),
      type: typeof j["type"] === "string" ? (j["type"] as string) : "",
    };
  });
}

function buildFeedback(
  fluency: number,
  pacing: number,
  clarity: number,
  pm: PauseMetrics,
  sm: SpeechStats,
): string[] {
  const items: string[] = [];
  items.push(
    `You spoke at ${Math.round(sm.speechRateWpm)} WPM` +
      (pm.pauseCount > 0
        ? ` with ${pm.pauseCount} pauses (avg ${pm.avgPauseDuration.toFixed(1)}s).`
        : "."),
  );
  if (fluency < 50) {
    items.push("Fluency needs attention — focus on reducing long silences.");
  } else if (pacing < 50) {
    items.push("Work on keeping a steadier speaking pace throughout.");
  } else if (clarity < 50) {
    items.push("Articulation of word endings could be sharper.");
  } else {
    items.push("Good overall delivery — keep practising to build consistency.");
  }
  return items;
}

/**
 * Parses both shapes the API can return:
 *   Full response  -> {"file": "...", "analysis": { ... }}
 *   Bare analysis  -> { "transcript": ..., "scores": ..., ... }
 */
export function parseAnalysisResult(json: unknown): AnalysisResult {
  const root = (json ?? {}) as Record<string, unknown>;
  const a = (root["analysis"] ?? root) as Record<string, unknown>;

  const rawScores = (a["scores"] ?? {}) as Record<string, unknown>;

  const fluencyScore = clampScore(rawScores["fluency"]);
  const pacingScore = clampScore(rawScores["pacing"]);
  const clarityScore = clampScore(rawScores["clarity"]);
  const confidenceScore = clampScore(rawScores["confidence"]);
  const vocabularyScore = clampScore(rawScores["vocabulary"]);
  const pronunciationScore = clampScore(rawScores["pronunciation"]);
  const overallScore = clampScore(rawScores["overall"]);

  const pausesMap = (a["pauses"] ?? {}) as Record<string, unknown>;
  const pm = parsePauseMetrics((pausesMap["metrics"] ?? {}) as Record<string, unknown>);
  const sm = parseSpeechStats((a["speech"] ?? {}) as Record<string, unknown>);

  const rawFeedback = a["feedback"];
  const feedbackList: string[] =
    Array.isArray(rawFeedback) && rawFeedback.length > 0
      ? rawFeedback.map((e) => String(e))
      : buildFeedback(fluencyScore, pacingScore, clarityScore, pm, sm);

  // Real per-dimension potentials come from the backend (same formula re-run
  // with that dimension's penalty events removed). The old +12 heuristic only
  // survives as a fallback for responses that predate the field.
  const breakdowns = parseBreakdowns(rawScores["breakdown"]);
  const potentialOf = (category: MetricCategory, score: number): number =>
    breakdowns[category]?.potential ?? potentialFor(score);

  const rawOverallPotential = rawScores["overall_potential"];
  const overallPotential =
    typeof rawOverallPotential === "number" && rawOverallPotential > 0
      ? clampScore(rawOverallPotential)
      : potentialFor(overallScore);

  return {
    overallScore,
    overallPotential,
    fluency: { score: fluencyScore, potential: potentialOf("Fluency", fluencyScore) },
    pacing: { score: pacingScore, potential: potentialOf("Pacing", pacingScore) },
    clarity: { score: clarityScore, potential: potentialOf("Clarity", clarityScore) },
    confidence: { score: confidenceScore, potential: potentialOf("Confidence", confidenceScore) },
    vocabulary: { score: vocabularyScore, potential: potentialOf("Vocabulary", vocabularyScore) },
    pronunciation: { score: pronunciationScore, potential: potentialFor(pronunciationScore) },
    feedback: feedbackList,
    transcript: typeof a["transcript"] === "string" ? (a["transcript"] as string) : "",
    segments: parseSegments(a["segments"]),
    duration: toNumber(a["duration"]),
    pauseMetrics: pm,
    speechMetrics: sm,
    timeline: parseTimeline(a["timeline"]),
    idealRanges: parseIdealRanges(rawScores["ideal_ranges"]),
    breakdowns,
  };
}

/** Silence segments as 0-1 waveform highlight ranges. */
export function getSilenceHighlights(result: AnalysisResult): Array<{ start: number; end: number }> {
  if (result.duration <= 0) return [];
  return result.timeline
    .filter((s) => s.type === "silence")
    .map((s) => ({
      start: Math.min(1, Math.max(0, s.start / result.duration)),
      end: Math.min(1, Math.max(0, s.end / result.duration)),
    }));
}

/** The lowest-scoring implemented metric (for the coaching card). Pronunciation excluded. */
export function getWorstMetric(result: AnalysisResult): MetricCategory {
  const scores: Array<[MetricCategory, number]> = [
    ["Fluency", result.fluency.score],
    ["Pacing", result.pacing.score],
    ["Clarity", result.clarity.score],
    ["Confidence", result.confidence.score],
    ["Vocabulary", result.vocabulary.score],
  ];
  return scores.reduce((a, b) => (a[1] <= b[1] ? a : b))[0];
}
